### ISSUE TITLE
How to annotate an owner properly?

### ISSUE THREAD

**Issue by brevzin** [mgehre/llvm-project#97]
Here's a simple example:

```cpp
#include <vector>
#include <string>

void use(char);
auto f() -> std::vector<std::string>;

int main() {
    for (char c : f()[1]) {
        use(c);
    }
}
```

This, correctly and awesomely, issues a warning on the for loop because of the object being destroyed. Cool.

Now let's say I have my own custom vector of strings, which I'm calling `VS` because I'm not creative:

```cpp
#include <vector>
#include <string>

void use(char);

struct [[gsl::Owner]] VS {
    auto operator[](size_t i) -> std::string&;
    auto begin() -> std::vector<std::string>::iterator;
    auto end() -> std::vector<std::string>::iterator;

    std::vector<std::string> v;
};

auto f() -> VS;

int main() {
    for (char c : f()[1]) {
        use(c);
    }
}
```

This does _not_ warn, even though it's basically the same example and I tried to annotate `VS` as being an owner. Is there some way to still get a warning for such code?

[Obligatory compiler explorer link](https://godbolt.org/z/ccM6G1).

**Comment by mgehre** [comment]
This seems to be an issue with determining the pointee type of the `Owner` VS.
Adding some debugging and an extra member makes it correctly deduce the pointee type, see https://godbolt.org/z/Erb77o

I thought that ` [[gsl::Owner(std::string)]]` was a way to manually specify that, but it didn't work.
Also declaring `operator[]` is supposed to be enough to correctly determine the pointee type.

I'll leave this bug open until that is fixed.

**Comment by brevzin** [comment]
> Also declaring `operator[]` is supposed to be enough to correctly determine the pointee type.

I think if the type is a range (has `begin`/`end`), you should be able to use that too?

**Comment by mgehre** [comment]
Yes, it *should*. The respective code is in https://github.com/mgehre/llvm-project/blob/lifetime/clang/lib/Analysis/LifetimeTypeCategory.cpp#L332

**Comment by mgehre** [comment]
For some reason, it doesn't recognize `std::vector<std::string>::iterator`   as Pointer.

**Comment by mgehre** [comment]
It seems to be an issue when `std::vector<std::string>::begin` is not used and thus not instantiated, see
https://godbolt.org/z/6bbqKT
